build-trust/ockam
# fix(rust): retrieve the correct key id for a public key with aws kms

This PR fixes the use of the AWS KMS as a backend for a `Vault`.

# Observed issue

When using the `AttachKey` command to create an identity a call to the AWS KMS client was performed to retrieve the key id corresponding to a given public key. 

This call is necessary because when we compute an `IdentityIdentifier` for an identity we derive it from the key id related to the public key used for the first change in the change history.
Before this PR the call to `kms_client.compute_key_id(&public_key)` was not implemented correctly and could not retrieve the key id.

# Diagnostic

The trouble is that there is no efficient implementation of this operation with the AWS KMS. That operation needs to list all the existing keys to retrieve the key with the right public key.

# Proposed fix

This PR fixes the issue by:

 - adding a local storage for public key <-> key id when the AWS security module is used
 - retrieving the key id directly from storage when the public key id is given and the mapping exists
 - retrieving the key id from the AWS KMS if it has never been accessed before (and then store the mapping locally)
 - doing a local verification of signatures instead of using the AWS KMS and incurring a remote call
 - removing the computation of an `IdentityIdentifier` from the key id associated to a public key. Now the identity identifier is directly a sha256 hash of the first public key of an identity

# Tests

Several tests have been added to:

 - check the low-level behaviour of the storage of the public key <-> key id mapping
 - check that a signature can be verified both locally and remotely with a given AWS KMS key
 - check that the identity creation done when running the `attach-key` command works when using an AWS KMS backend

Note: those tests cannot be part of the test suite for now because they require an access to an AWS account to create keys.
